src/atom_utils.F

```diff
@@ -1477,172 +1558,290 @@ SUBROUTINE potential_analytic(cpot, la, lb, nu, basis)
+!> \brief This routine calculates the longrange Coulomb potential of a product of two Gaussian with given angular momentum
+!> \param erfa Array will contain the potential
+!> \param a Exponent of first Gaussian charge distribution
+!> \param b Exponent of second Gaussian charge distribution
+!> \param ll Sum of angular momentum quantum numbers building one charge distribution
+!> \param nu Angular momentum of interaction, (ll-nu) should be even.
+!> \param rad Radial grid
+!> \param omega Range-separation parameter
+! **************************************************************************************************
+   PURE SUBROUTINE potential_longrange_analytic(erfa, a, b, ll, nu, rad, omega)
+      REAL(KIND=dp), DIMENSION(:), INTENT(OUT)           :: erfa
+      REAL(KIND=dp), INTENT(IN)                          :: a, b
+      INTEGER, INTENT(IN)                                :: ll, nu
+      REAL(KIND=dp), DIMENSION(:), INTENT(IN)            :: rad
+      REAL(KIND=dp), INTENT(IN)                          :: omega
+
+      CHARACTER(len=*), PARAMETER :: routineN = 'potential_longrange_analytic', &
+         routineP = moduleN//':'//routineN
+
+      INTEGER                                            :: i, lambda
+      REAL(KIND=dp)                                      :: ab, oab, pab, prel, sab
+      REAL(KIND=dp), ALLOCATABLE, DIMENSION(:)           :: expa, z
+
+      IF (MOD(ll - nu, 2) == 0 .AND. ll >= nu .AND. nu >= 0) THEN
+         ALLOCATE (expa, SOURCE=rad)
+         ALLOCATE (z, SOURCE=rad)
+
+         lambda = (ll - nu)/2
+         ab = a + b
+         sab = SQRT(ab)
+         pab = omega**2*ab/(omega**2 + ab)
+         prel = pab/ab
+         z(:) = SQRT(pab)*rad(:)
+         oab = fac(lambda)/SQRT(ab)**(ll + 2)/4.0_dp*SQRT(prel)**(nu + 1)*(2.0_dp*REAL(nu, KIND=dp) + 1.0_dp)
+         expa(:) = EXP(-z(:)**2)
+         lambda = (ll - nu)/2
+
+         IF (lambda > 0) THEN
+            erfa = 0.0_dp
+            DO i = 1, lambda
+               erfa = erfa + (-prel)**i/REAL(i, KIND=dp)*binomial_gen(lambda + nu + 0.5_dp, lambda - i)* &
+                      assoc_laguerre(z, REAL(nu, KIND=dp) + 0.5_dp, i - 1)
+            END DO
+            erfa = erfa*expa*z**nu
+
+            erfa = erfa + 2.0_dp*binomial_gen(lambda + nu + 0.5_dp, lambda)*znFn(z, expa, nu)
+         ELSE
+            erfa = 2.0_dp*znFn(z, expa, nu)
+         END IF
+
+         erfa = erfa*oab
+      ELSE
+         erfa = 0.0_dp
+      END IF
+
+      DEALLOCATE (expa, z)
+
+   END SUBROUTINE potential_longrange_analytic
+
+! **************************************************************************************************
+!> \brief Boys' function times z**n
+!> \param z ...
+!> \param expa ...
+!> \param n ...
+!> \return ...
+! **************************************************************************************************
+   ELEMENTAL FUNCTION znFn(z, expa, n) RESULT(res)
```

Comment:
`PURE` maybe?

Reply:
well, I realized that `ELEMENTAL` already implies `PURE` and there's the `IMPURE` instead